### ISSUE TITLE
"Frozen" chunks after first maxConcurrentParts are done

### ISSUE THREAD

**Issue by jakubzitny** [TTLabs/EvaporateJS#227]
Hi, this is possibly related to #223 but I am not sure. We're having problems uploading files larger than `x` MB (where `x` is somewhere around `maxConcurrentParts * partSize`). When Evaporate reaches this number it is just reporting the uploaded size in `progress` callback, but it is not moving anywhere.

Symptoms I have been able to detect: Let's say the `maxConcurrentParts=10` and the file is larger than `10*partSize`. The whole file is split into 10+ (let's say 20) different parts and the first 10 will start evaporating. When these are successfully done, the 10 remaining parts do not start, `loadedBytes` stay at `0`, `status=2` and `attempts=1`.

![screen shot 2016-09-01 at 16 36 13](https://cloud.githubusercontent.com/assets/3315662/18171533/b0ab6f7e-7062-11e6-8375-e638e0f7632a.png)

Any ideas what's up with that? What other info should I provide?

I am using the latest Evaporate, custom `signResponseHandler` where the requests are signed by temporary `STS` credentials (likely unrelated) and everything works properly for small files (small being smaller than `maxConcurrentParts * partSize`).

**Comment by bikeath1337** [comment]
If feasible, try branch `retry_issue`, and see if it is addressed. Many thanks.

**Comment by jakubzitny** [comment]
Yeah I just tried that one too and it's doing the same.

**Comment by fkjaekel** [comment]
I tried 1.5.5 today, there is a 213.4MB file that always hangs at 14.7%.

I tried the retry_issue branch, but I'm getting this error: `Error listing parts for getUploadParts() starting at part # 0`

I'm able to upload this file using 1.4.0. I'll attach logs.

[logs.zip](https://github.com/TTLabs/EvaporateJS/files/450362/logs.zip)

**Comment by jakubzitny** [comment]
@fkjaekel I am not 100% sure that's related. I think I used to get that error when I was e.g. uploading the same file to the same s3 path. Can you maybe try it one more time and send me the logs but with added one console.log? I added console log to evaporate's `processPartsList()` in the for loop.

```
console.log('dd processing part', i, part) // the 'dd' is for filtering in browser
```

Also I was not using `computeContentMd5` and `cryptoMd5Method`. I added it and it stopped happening for a while. But I just tried it again with the Md5s and it happened again :\

@bikeath1337 can you maybe confirm that having `computeContentMd5=false` can cause similar problems, please?

**Comment by bikeath1337** [comment]
@jakubzitny using computeContentMd5 is something you ought to be doing. I fact, everyone ought ot be using it. Otherwise, you can't be guaranteed that your file was copied with complete fidelity. If you had issues with the MD5 checksum, it is because AWS was doing its job making sure that the checksums match.

In other words, if there is a retry issue, you would see it when using checksums in the cases of the transmission was corrupted.

Thus: keep the md5. Let's address why the retry logic isn't working.

**Comment by bikeath1337** [comment]
I'm looking at this now.

**Comment by jakubzitny** [comment]
Thanks for the clarification.

Back to the main issue. I think the problem I am having arose in the fix for issue #205. When I checkout to commit e27854a7a098e0e5e8be2c6359180d2989b07a99 everything works fine. And on the next one — ffec664c00f4f680ef76750611685e1b5ef83098 — it is doing exactly what I described above.

I just tried it several times with the example code from `example/evaporate_example.html` and `example/signing_example.py` with my bucket just to be sure it has nothing to do with my usage of Evaporate.

**Comment by jakubzitny** [comment]
[This line](https://github.com/TTLabs/EvaporateJS/commit/ffec664c00f4f680ef76750611685e1b5ef83098#diff-f7216cb05b1b8666b24de80f3d918b24R679) in particular.

If I remove the first part of the condition `evaporatingCount < con.maxConcurrentParts && ...` it seems to be working again.

**Comment by bikeath1337** [comment]
I restored a line a removed from Tremby that was causing the issue. I pushed it to TTLabs/retry_issue now. Give it a try.

https://github.com/TTLabs/EvaporateJS/commit/6898c0ddc2a8bd2f306617f6f789afe60d28ffa6

**Comment by bikeath1337** [comment]
By removing the condition in https://github.com/TTLabs/EvaporateJS/commit/ffec664c00f4f680ef76750611685e1b5ef83098#diff-f7216cb05b1b8666b24de80f3d918b24R679, it introduces the inability to stay within the maxConcurrentParts.

**Comment by jakubzitny** [comment]
Yes @bikeath1337 6898c0d seems to be helping. Latest `retry_issue` works correctly.

Thanks for prompt fix 🙇

**Comment by bikeath1337** [comment]
This should be resoled in #236 . Was able to reproduce and resolve.
